Index: swgfs-1.8.1.1_client/swgfs/llite/llite_mmap.c
===================================================================
--- swgfs-1.8.1.1_client.orig/swgfs/llite/llite_mmap.c	2011-01-08 01:09:54.000000000 +0800
+++ swgfs-1.8.1.1_client/swgfs/llite/llite_mmap.c	2011-11-22 01:45:44.000000000 +0800
@@ -815,3 +815,11 @@
         RETURN(rc);
 }
 #endif
+
+#ifndef LIXI_20111121
+struct vm_operations_struct *ll_hrfs_get_vm_ops(void)
+{
+        return &ll_file_vm_ops;
+}
+EXPORT_SYMBOL(ll_hrfs_get_vm_ops);
+#endif /* LIXI_20111121 */
