Index: swgfs-1.8.1.1/swgfs/llite/file.c
===================================================================
--- swgfs-1.8.1.1.orig/swgfs/llite/file.c	2011-05-19 08:02:59.000000000 +0800
+++ swgfs-1.8.1.1/swgfs/llite/file.c	2011-05-19 08:03:06.000000000 +0800
@@ -5370,3 +5370,43 @@
 EXPORT_SYMBOL(ll_clear_parent_inode);
 #endif /* LIXI_20110331 */
 
+#ifndef LIXI_20110518
+int swgfs_getstripe(struct dentry *dentry, struct lov_user_md *lump)
+{
+	struct inode *inode = dentry->d_inode;
+	struct lov_mds_md *lmm = NULL;
+	struct ptlrpc_request *request = NULL;
+	int rc = 0, lmmsize = 0;
+
+
+	LASSERT(inode);
+
+	if (S_ISREG(inode->i_mode)) {
+			rc = ll_lov_getstripe_ea_info(dentry->d_parent->d_inode,
+										  dentry->d_name.name, &lmm,
+										  &lmmsize, &request);
+	} else if (S_ISDIR(inode->i_mode)) {
+			rc = ll_dir_getstripe(inode, &lmm, &lmmsize, &request);
+	} else {
+			rc = -ENODATA;
+	}
+
+	if (rc < 0) {
+		GOTO(out, rc);
+	}
+
+	if (sizeof(*lump) < lmmsize) {
+			CERROR("server bug: replied size %d > %d for %s\n",
+				   lmmsize, (int)sizeof(*lump), dentry->d_name.name);
+			GOTO(out, rc = -ERANGE);
+	}
+  	memcpy(lump, lmm, lmmsize);
+
+	rc = lmmsize;
+out:
+	ptlrpc_req_finished(request);
+	return rc;
+}
+EXPORT_SYMBOL(swgfs_getstripe);
+#endif /* LIXI_20110518 */
+
