Index: lustre-2.1.52/lustre/llite/llite_mmap.c
===================================================================
--- lustre-2.1.52.orig/lustre/llite/llite_mmap.c	2012-03-28 14:19:00.000000000 +0800
+++ lustre-2.1.52/lustre/llite/llite_mmap.c	2012-03-28 14:19:10.000000000 +0800
@@ -440,3 +440,11 @@
 
         RETURN(rc);
 }
+
+#ifndef DISABLE_MTFS_SUPPORT
+struct vm_operations_struct *mtfs_ll_get_vm_ops(void)
+{
+        return &ll_file_vm_ops;
+}
+EXPORT_SYMBOL(mtfs_ll_get_vm_ops);
+#endif /* !DISABLE_MTFS_SUPPORT */
