Index: lustre-2.1.52/lustre/include/lustre/lustre_idl.h
===================================================================
--- lustre-2.1.52.orig/lustre/include/lustre/lustre_idl.h	2012-03-28 14:18:57.000000000 +0800
+++ lustre-2.1.52/lustre/include/lustre/lustre_idl.h	2012-03-28 14:19:13.000000000 +0800
@@ -1335,7 +1335,6 @@
 #define XATTR_NAME_LMV          "trusted.lmv"
 #define XATTR_NAME_LINK         "trusted.link"
 
-
 struct lov_mds_md_v3 {            /* LOV EA mds/wire data (little-endian) */
         __u32 lmm_magic;          /* magic number = LOV_MAGIC_V3 */
         __u32 lmm_pattern;        /* LOV_PATTERN_RAID0, LOV_PATTERN_RAID1 */
@@ -1403,11 +1402,22 @@
 #define OBD_MD_FLRMTRSETFACL (0x0004000000000000ULL) /* lfs rsetfacl case */
 #define OBD_MD_FLRMTRGETFACL (0x0008000000000000ULL) /* lfs rgetfacl case */
 
+#ifndef DISABLE_MTFS_SUPPORT
+#define OBD_MD_FLMTFSFLAG    (0x0010000000000000ULL) /* mtfs flag */
+/* d_revalidate should update mtfs flag */
+#define OBD_MD_FLGETATTR (OBD_MD_FLID    | OBD_MD_FLATIME | OBD_MD_FLMTIME | \
+                          OBD_MD_FLCTIME | OBD_MD_FLSIZE  | OBD_MD_FLBLKSZ | \
+                          OBD_MD_FLMODE  | OBD_MD_FLTYPE  | OBD_MD_FLUID   | \
+                          OBD_MD_FLGID   | OBD_MD_FLFLAGS | OBD_MD_FLNLINK | \
+                          OBD_MD_FLGENER | OBD_MD_FLRDEV  | OBD_MD_FLGROUP | \
+                          OBD_MD_FLMTFSFLAG)
+#else /* !DISABLE_MTFS_SUPPORT */
 #define OBD_MD_FLGETATTR (OBD_MD_FLID    | OBD_MD_FLATIME | OBD_MD_FLMTIME | \
                           OBD_MD_FLCTIME | OBD_MD_FLSIZE  | OBD_MD_FLBLKSZ | \
                           OBD_MD_FLMODE  | OBD_MD_FLTYPE  | OBD_MD_FLUID   | \
                           OBD_MD_FLGID   | OBD_MD_FLFLAGS | OBD_MD_FLNLINK | \
                           OBD_MD_FLGENER | OBD_MD_FLRDEV  | OBD_MD_FLGROUP)
+#endif /* DISABLE_MTFS_SUPPORT */
 
 /* don't forget obdo_fid which is way down at the bottom so it can
  * come after the definition of llog_cookie */
@@ -1651,6 +1661,14 @@
         __u64          padding_10;
 }; /* 216 */
 
+#ifndef DISABLE_MTFS_SUPPORT
+/*
+ * Use field padding_5 of struct mdt_body,
+ * without breaking wirecheck.
+ */
+#define MTFS_MDT_BODY_FLAG padding_5
+#endif /* !DISABLE_MTFS_SUPPORT */
+
 extern void lustre_swab_mdt_body (struct mdt_body *b);
 
 struct mdt_ioepoch {
@@ -1784,6 +1802,14 @@
         __u32           sa_padding_5;
 };
 
+#ifndef DISABLE_MTFS_SUPPORT
+/*
+ * Use field sa_padding_2 of struct mdt_rec_setattr,
+ * without breaking wirecheck.
+ */
+#define sa_mtfs_flag sa_padding_2
+#endif /* !DISABLE_MTFS_SUPPORT */
+
 extern void lustre_swab_mdt_rec_setattr (struct mdt_rec_setattr *sa);
 
 /*
@@ -1851,6 +1877,14 @@
 #define MDS_OPEN_NORESTORE  0100000000000ULL /* Do not restore file at open */
 #define MDS_OPEN_NEWSTRIPE  0200000000000ULL /* New stripe needed (restripe or
                                               * hsm restore) */
+#ifndef DISABLE_MTFS_SUPPORT
+/*
+ * All 32th bits exhausted, use an unused bit.
+ */
+
+/* set when modifying dir's mtfs_flag at mds */
+#define MTFS_MDT_SET_FLAG	MDS_OPEN_JOIN_FILE
+#endif /* !DISABLE_MTFS_SUPPORT */
 
 /* permission for create non-directory file */
 #define MAY_CREATE      (1 << 7)
Index: lustre-2.1.52/lustre/llite/llite_internal.h
===================================================================
--- lustre-2.1.52.orig/lustre/llite/llite_internal.h	2012-03-28 14:18:57.000000000 +0800
+++ lustre-2.1.52/lustre/llite/llite_internal.h	2012-03-28 14:19:13.000000000 +0800
@@ -126,6 +126,12 @@
 
 };
 
+#ifndef DISABLE_MTFS_SUPPORT
+struct mtfs_ll_inode_info {
+        __u32 mtfs_flag;
+};
+#endif /* !DISABLE_MTFS_SUPPORT */
+
 struct ll_inode_info {
         int                     lli_inode_magic;
         cfs_semaphore_t         lli_size_sem;           /* protect open and change size */
@@ -210,6 +216,9 @@
          * serialize normal readdir and statahead-readdir
          */
         cfs_semaphore_t         lli_readdir_sem;
+#ifndef DISABLE_MTFS_SUPPORT
+        struct mtfs_ll_inode_info lli_mtfs_info;
+#endif /* !DISABLE_MTFS_SUPPORT */
 };
 
 /*
Index: lustre-2.1.52/lustre/llite/llite_lib.c
===================================================================
--- lustre-2.1.52.orig/lustre/llite/llite_lib.c	2012-03-28 14:18:57.000000000 +0800
+++ lustre-2.1.52/lustre/llite/llite_lib.c	2012-03-28 14:19:13.000000000 +0800
@@ -1545,6 +1545,9 @@
         cfs_up(&lli->lli_size_sem);
 }
 
+#ifndef DISABLE_MTFS_SUPPORT
+#include <mtfs/mtfs_flag.h>
+#endif /* !DISABLE_MTFS_SUPPORT */
 void ll_update_inode(struct inode *inode, struct lustre_md *md)
 {
         struct ll_inode_info *lli = ll_i2info(inode);
@@ -1716,6 +1719,19 @@
                 LASSERT(md->oss_capa);
                 ll_add_capa(inode, md->oss_capa);
         }
+#ifndef DISABLE_MTFS_SUPPORT
+        if (body->valid & OBD_MD_FLMTFSFLAG) {
+                __u32 mtfs_flag = body->MTFS_MDT_BODY_FLAG;
+                struct ll_inode_info *lli = ll_i2info(inode);
+                struct mtfs_ll_inode_info *mtfs_info = &(lli->lli_mtfs_info);
+
+                if (unlikely(!mtfs_flag_is_valid(mtfs_flag))) {
+                        MERROR("mtfs_flag 0x%x is not valid\n", mtfs_flag);
+                }
+                mtfs_info->mtfs_flag = mtfs_flag;
+                MDEBUG("updating inode's mtfs_flag to 0x%x\n", mtfs_flag);
+        }
+#endif /* !DISABLE_MTFS_SUPPORT */
 }
 
 void ll_read_inode2(struct inode *inode, void *opaque)
@@ -2228,3 +2244,108 @@
 
         RETURN(0);
 }
+
+#ifndef DISABLE_MTFS_SUPPORT
+int mtfs_ll_inode_get_flag(struct inode *inode, __u32 *mtfs_flag)
+{
+        int rc = 0;
+        struct ll_sb_info *sbi = ll_i2sbi(inode);
+        struct ptlrpc_request *req = NULL;
+        struct mdt_body *body = NULL;
+        struct md_op_data *op_data = NULL;
+        __u32 flag = 0;
+
+        op_data = ll_prep_md_op_data(NULL, inode, NULL, NULL,
+                                     0, 0, LUSTRE_OPC_ANY,
+                                     NULL);
+        if (op_data == NULL) {
+                rc = -ENOMEM;
+                goto out;
+        }
+
+        op_data->op_valid = OBD_MD_FLMTFSFLAG;
+
+        rc = md_getattr(sbi->ll_md_exp, op_data, &req);
+        ll_finish_md_op_data(op_data);
+        if (rc) {
+                MERROR("failed to get flag of inode %lu: %d\n",
+                       inode->i_ino, rc);
+                goto out;
+        }
+
+        body = req_capsule_server_get(&req->rq_pill, &RMF_MDT_BODY);
+        LASSERT(body != NULL);
+
+        if (body->valid & OBD_MD_FLMTFSFLAG) {
+                flag = body->MTFS_MDT_BODY_FLAG;
+                if (unlikely(!mtfs_flag_is_valid(flag))) {
+                        MERROR("get an error flag of inode %lu\n",
+                               inode->i_ino);
+                        rc = -EPERM;
+                        goto out;
+                }
+        } else {
+                MERROR("failed to get flag of inode %lu: no data returned\n",
+                       inode->i_ino);
+                rc = -ENODATA;
+                goto out;
+        }
+
+        *mtfs_flag = flag;
+
+        /* Update flag of inode */
+        rc = ll_prep_inode(&inode, req, NULL);
+
+ out:
+        if (req) {
+                ptlrpc_req_finished(req);
+        }
+        return rc;
+}
+EXPORT_SYMBOL(mtfs_ll_inode_get_flag);
+
+int mtfs_ll_inode_set_flag(struct inode *inode, __u32 mtfs_flag)
+{
+        struct ll_inode_info *lli = ll_i2info(inode);
+        struct md_op_data *op_data = NULL;
+        struct md_open_data *mod = NULL;
+        int rc = 0, rc1 = 0;
+        MENTRY();
+
+        if (unlikely(!mtfs_flag_is_valid(mtfs_flag))) {
+                rc = -EPERM;
+                goto out;
+        }
+
+        ll_stats_ops_tally(ll_i2sbi(inode), LPROC_LL_SETATTR, 1);
+
+        OBD_ALLOC_PTR(op_data);
+        if (op_data == NULL) {
+                rc = -ENOMEM;
+                goto out;
+        }
+
+        /* Seems no need to lock lli->lli_trunc_sem */
+
+        op_data->op_attr.ia_valid |= MTFS_MDT_SET_FLAG;
+        op_data->op_mtfs_flag = mtfs_flag;
+
+        CERROR("in %x, %u\n",
+               op_data->op_attr.ia_valid, op_data->op_mtfs_flag);
+        rc = ll_md_setattr(inode, op_data, &mod);
+        if (rc) {
+                goto out;
+        }
+
+        ll_ioepoch_open(lli, op_data->op_ioepoch);
+out:
+        if (op_data) {
+                if (op_data->op_ioepoch) {
+                        rc1 = ll_setattr_done_writing(inode, op_data, mod);
+                }
+                ll_finish_md_op_data(op_data);
+        }
+        MRETURN(rc);
+}
+EXPORT_SYMBOL(mtfs_ll_inode_set_flag);
+#endif /* !DISABLE_MTFS_SUPPORT */
Index: lustre-2.1.52/lustre/include/lu_object.h
===================================================================
--- lustre-2.1.52.orig/lustre/include/lu_object.h	2012-03-28 14:18:57.000000000 +0800
+++ lustre-2.1.52/lustre/include/lu_object.h	2012-03-28 14:19:13.000000000 +0800
@@ -437,6 +437,10 @@
         __u32          la_blksize;
         /** real device */
         __u32          la_rdev;
+#ifndef DISABLE_MTFS_SUPPORT
+        /** mtfs flag */
+        __u32          la_mtfs_disk_flag;
+#endif /* !DISABLE_MTFS_SUPPORT */
         /**
          * valid bits
          *
Index: lustre-2.1.52/lustre/mdt/mdt_handler.c
===================================================================
--- lustre-2.1.52.orig/lustre/mdt/mdt_handler.c	2012-03-28 14:18:57.000000000 +0800
+++ lustre-2.1.52/lustre/mdt/mdt_handler.c	2012-03-28 14:19:13.000000000 +0800
@@ -405,6 +405,11 @@
         b->blocks = ma->ma_som->msd_blocks;
 }
 
+#ifndef DISABLE_MTFS_SUPPORT
+#include <mtfs/debug.h>
+#include <mtfs/mtfs_flag.h>
+#endif /* !DISABLE_MTFS_SUPPORT */
+
 void mdt_pack_attr2body(struct mdt_thread_info *info, struct mdt_body *b,
                         const struct lu_attr *attr, const struct lu_fid *fid)
 {
@@ -460,6 +465,13 @@
         if (b->valid & OBD_MD_FLSIZE)
                 CDEBUG(D_VFSTRACE, DFID": returning size %llu\n",
                        PFID(fid), (unsigned long long)b->size);
+#ifndef DISABLE_MTFS_SUPPORT
+        if (attr->la_mtfs_disk_flag != 0) {
+                MASSERT(mtfs_disk_flag_is_valid(attr->la_mtfs_disk_flag));
+                b->MTFS_MDT_BODY_FLAG = attr->la_mtfs_disk_flag & MTFS_FLAG_DISK_MASK;
+                b->valid |= OBD_MD_FLMTFSFLAG;
+        }
+#endif /* !DISABLE_MTFS_SUPPORT */
 }
 
 static inline int mdt_body_has_lov(const struct lu_attr *la,
@@ -525,6 +537,13 @@
         if (ma->ma_need & MA_SOM)
                 ma->ma_som = &info->mti_u.som.data;
 
+#ifndef DISABLE_MTFS_SUPPORT
+        if (reqbody->valid & OBD_MD_FLMTFSFLAG) {
+                MDEBUG("getting flag\n");
+                ma->ma_need |= MA_MTFS;
+        }
+#endif /* DISABLE_MTFS_SUPPORT */
+
         rc = mo_attr_get(env, next, ma);
         if (unlikely(rc)) {
                 CERROR("getattr error for "DFID": %d\n",
Index: lustre-2.1.52/lustre/mdd/mdd_object.c
===================================================================
--- lustre-2.1.52.orig/lustre/mdd/mdd_object.c	2012-03-28 14:18:57.000000000 +0800
+++ lustre-2.1.52/lustre/mdd/mdd_object.c	2012-03-28 14:19:13.000000000 +0800
@@ -751,6 +751,50 @@
         RETURN(0);
 }
 
+#ifndef DISABLE_MTFS_SUPPORT
+#include <mtfs/debug.h>
+#include <mtfs/mtfs_flag.h>
+
+int mtfs_mdd_get_flag(const struct lu_env *env, struct mdd_object *mdd_obj,
+                       struct md_attr *ma)
+{
+        int rc = 0;
+        __u32 disk_flag = 0;
+        int flag_size = sizeof(disk_flag);
+        MENTRY();
+
+        rc = mdd_get_md(env, mdd_obj, &disk_flag, &flag_size,
+                        XATTR_NAME_MTFS_FLAG);
+        if (unlikely(rc < 0)) {
+                MERROR("failed to get mtfs flag, rc = %d, "
+                "flag_size = %d\n", rc, flag_size);
+                ma->ma_attr.la_mtfs_disk_flag = 0;
+                goto out;
+        } else if (rc == 0) {
+        	MDEBUG("flag is not set\n");
+        	ma->ma_attr.la_mtfs_disk_flag = MTFS_FLAG_DISK_SYMBOL;
+        	goto out;
+        } else if (flag_size != sizeof(disk_flag)) {
+        	MDEBUG("got an invalid size of disk flag,"
+        	       " expect %ld, got %d\n",
+        	       sizeof(disk_flag), flag_size);
+        	ma->ma_attr.la_mtfs_disk_flag = 0;
+        	goto out;
+        }
+
+        if (unlikely(!mtfs_disk_flag_is_valid(disk_flag))) {
+                MERROR("got an invalid disk flag\n");
+                ma->ma_attr.la_mtfs_disk_flag = 0;
+                goto out;
+        }
+
+        ma->ma_attr.la_mtfs_disk_flag = disk_flag;
+out:
+        rc = 0;
+        MRETURN(rc);
+}
+#endif /* !DISABLE_MTFS_SUPPORT */
+
 int mdd_attr_get_internal(const struct lu_env *env, struct mdd_object *mdd_obj,
                                  struct md_attr *ma)
 {
@@ -783,6 +827,12 @@
                         rc = mdd_def_acl_get(env, mdd_obj, ma);
         }
 #endif
+
+#ifndef DISABLE_MTFS_SUPPORT
+        if (rc == 0 && ma->ma_need & MA_MTFS) {
+                rc = mtfs_mdd_get_flag(env, mdd_obj, ma);
+        }
+#endif /* DISABLE_MTFS_SUPPORT */
         CDEBUG(D_INODE, "after getattr rc = %d, ma_valid = "LPX64" ma_lmm=%p\n",
                rc, ma->ma_valid, ma->ma_lmm);
         RETURN(rc);
@@ -792,8 +842,13 @@
                                  struct mdd_object *mdd_obj, struct md_attr *ma)
 {
         int rc;
+#ifndef DISABLE_MTFS_SUPPORT
+        int needlock = ma->ma_need &
+                       (MA_LOV | MA_LMV | MA_ACL_DEF | MA_HSM | MA_SOM | MA_PFID | MA_MTFS);
+#else
         int needlock = ma->ma_need &
                        (MA_LOV | MA_LMV | MA_ACL_DEF | MA_HSM | MA_SOM | MA_PFID);
+#endif
 
         if (needlock)
                 mdd_read_lock(env, mdd_obj, MOR_TGT_CHILD);
@@ -1449,6 +1504,24 @@
                                         md2mdd_obj(obj), handle);
 }
 
+#ifndef DISABLE_MTFS_SUPPORT
+static int mtfs_mdd_flag_set_md(const struct lu_env *env,
+                                struct mdd_object *mdd_obj,
+                                const struct md_attr *ma,
+                                struct thandle *handle)
+{
+        int             rc = 0;
+        struct lu_buf  *buf = NULL;
+        struct lu_attr *la = &((struct md_attr *)ma)->ma_attr;
+        MENTRY();
+
+        MASSERT(mtfs_disk_flag_is_valid(la->la_mtfs_disk_flag));
+        buf = mdd_buf_get(env, &la->la_mtfs_disk_flag, sizeof(la->la_mtfs_disk_flag));
+        rc = mdd_xattr_set_txn(env, mdd_obj, buf, XATTR_NAME_MTFS_FLAG, 0, handle);
+        MRETURN(rc);
+}
+#endif /* !DISABLE_MTFS_SUPPORT */
+
 /* set attr and LOV EA at once, return updated attr */
 static int mdd_attr_set(const struct lu_env *env, struct md_object *obj,
                         const struct md_attr *ma)
@@ -1583,6 +1656,13 @@
                         rc = mdd_lma_set_locked(env, mdd_obj, ma, handle);
 
         }
+
+#ifndef DISABLE_MTFS_SUPPORT
+        if (rc == 0 && ma->ma_valid & MA_MTFS) {
+                rc = mtfs_mdd_flag_set_md(env, mdd_obj, ma, handle);
+        }
+#endif /* !DISABLE_MTFS_SUPPORT */
+
 cleanup:
         if (rc == 0)
                 rc = mdd_attr_set_changelog(env, obj, handle,
Index: lustre-2.1.52/lustre/include/md_object.h
===================================================================
--- lustre-2.1.52.orig/lustre/include/md_object.h	2012-03-28 14:18:57.000000000 +0800
+++ lustre-2.1.52/lustre/include/md_object.h	2012-03-28 14:19:13.000000000 +0800
@@ -130,7 +130,13 @@
 /* (Layout lock will used #7 here) */
         MA_HSM       = (1 << 8),
         MA_SOM       = (1 << 9),
+#ifndef DISABLE_MTFS_SUPPORT
+        MA_PFID      = (1 << 10),
+        /* Need mtfs flag */
+        MA_MTFS      = (1 << 11)
+#else /* !DISABLE_MTFS_SUPPORT */
         MA_PFID      = (1 << 10)
+#endif /* DISABLE_MTFS_SUPPORT */
 };
 
 typedef enum {
Index: lustre-2.1.52/lustre/include/obd.h
===================================================================
--- lustre-2.1.52.orig/lustre/include/obd.h	2012-03-28 14:18:57.000000000 +0800
+++ lustre-2.1.52/lustre/include/obd.h	2012-03-28 14:19:13.000000000 +0800
@@ -1240,6 +1240,10 @@
 
         /* Operation type */
         __u32                   op_opc;
+#ifndef DISABLE_MTFS_SUPPORT
+        /* MTFS flag seting */
+        __u32                   op_mtfs_flag;
+#endif /* !DISABLE_MTFS_SUPPORT */
 };
 
 struct md_enqueue_info;
Index: lustre-2.1.52/lustre/mdc/mdc_lib.c
===================================================================
--- lustre-2.1.52.orig/lustre/mdc/mdc_lib.c	2012-03-28 14:18:57.000000000 +0800
+++ lustre-2.1.52/lustre/mdc/mdc_lib.c	2012-03-28 14:19:13.000000000 +0800
@@ -310,6 +310,14 @@
                 rec->sa_suppgid = op_data->op_attr.ia_gid;
         else
                 rec->sa_suppgid = op_data->op_suppgids[0];
+
+#ifndef DISABLE_MTFS_SUPPORT
+        if (op_data->op_attr.ia_valid & MTFS_MDT_SET_FLAG) {
+                rec->sa_valid |= MTFS_MDT_SET_FLAG;
+                rec->sa_mtfs_flag = op_data->op_mtfs_flag;
+        }
+#endif /* !DISABLE_MTFS_SUPPORT */
+
 }
 
 static void mdc_ioepoch_pack(struct mdt_ioepoch *epoch,
Index: lustre-2.1.52/lustre/mdt/mdt_lib.c
===================================================================
--- lustre-2.1.52.orig/lustre/mdt/mdt_lib.c	2012-03-28 14:18:57.000000000 +0800
+++ lustre-2.1.52/lustre/mdt/mdt_lib.c	2012-03-28 14:19:13.000000000 +0800
@@ -766,6 +766,9 @@
 }
 /* unpacking */
 
+#ifndef DISABLE_MTFS_SUPPORT
+#include <mtfs/mtfs_flag.h>
+#endif /* !DISABLE_MTFS_SUPPORT */
 static int mdt_setattr_unpack_rec(struct mdt_thread_info *info)
 {
         struct md_ucred         *uc  = mdt_ucred(info);
@@ -800,6 +803,15 @@
         la->la_mtime = rec->sa_mtime;
         ma->ma_valid = MA_INODE;
 
+#ifndef DISABLE_MTFS_SUPPORT
+        if (rec->sa_valid & MTFS_MDT_SET_FLAG) {
+                CERROR("setting mtfs flag %u\n", rec->sa_mtfs_flag);
+                MASSERT(mtfs_flag_is_valid(rec->sa_mtfs_flag));
+                la->la_mtfs_disk_flag = rec->sa_mtfs_flag | MTFS_FLAG_DISK_SYMBOL;
+                ma->ma_valid |= MA_MTFS;
+        }
+#endif /* !DISABLE_MTFS_SUPPORT */
+
         if (req_capsule_get_size(pill, &RMF_CAPA1, RCL_CLIENT))
                 mdt_set_capainfo(info, 0, rr->rr_fid1,
                                  req_capsule_client_get(pill, &RMF_CAPA1));
